### ISSUE TITLE
[SignalR] Java client code 401 when starting connection after stopping it

### ISSUE THREAD

**Issue by kosancicivan** [aspnet/AspNetCore#8429]
I am building Android app for connecting to Azure SignalR Service with bearer token authentication. Because of battery and memory constraints in Android, I want to stop the connection to hub, whenever user goes away from application.. But when I do, and then start the connection again, I am getting "java.lang.RuntimeException: Unexpected status code returned from negotiate: 401"

So basically when I build the hubConnection object, I provide logic for authentication in Single.defer, like the documentation recommends. It goes like this:
```
HubConnection hubConnection = HubConnectionBuilder.create("YOUR HUB URL HERE")
    .withAccessTokenProvider(Single.defer(() -> {
       String token = retrofitController.getApi().rxgetSignalRToken(url)
                                .subscribeOn(Schedulers.io())
                                .blockingGet();
                        if (token == null)
                            return Single.just("");
                        return Single.just(token);
    })).build();
```
Then when I call hubConnection.start, it successfully connects to SignalR service. After I call hubConnection.stop, next time when I want to connect I get previously mention exception. With the help of debugger, I find out that when I call hubConnection.start for the first time, it successfully acquire token with accesTokenProvider Single<String> I passed trough HubConnectionBuilder, but after that it calls code on line 247 of HubConnection.java: `this.accessTokenProvider = Single.just(negotiateResponse.getAccessToken());` and the accessTokenProvider is changed to reference different Single<String> then I supplied through withAccessTokenProvider method of builder class. Is there any way to override this, or am I doing something wrong, because the flow in .Net client is that token is always provided from AccessTokenProvider set in the HubConnectionBuilder class.
I used 3.0.0-preview-19075-0444 and 3.0.0-preview3-19153-02 versions of Java client.

**Comment by mikaelm12** [comment]
So it looks like you are correct. When connecting to the Azure SignalR Service we take the access token and wrap that in a single and override the `accessTokenProvider`. That behavior is correct, what is incorrect is that we don't keep a reference to the original passed in `accessTokenProvider` so we can reset it when we restart the connection. 
Good find.

**Comment by halter73** [comment]
We're also considering backporting this.

**Comment by aleksandra-majchrzak** [comment]
I'm experiencing the same problem in my android app  using `'com.microsoft.signalr:signalr:1.0.0'`. After I start the connection and stop it, when I try to connect again I get 401. I understand this was fixed in the Java client, but is there a solution to this in the new library?

**Comment by BrennanConroy** [comment]
This should be patched in version 1.0.5, try that version out.

**Comment by aleksandra-majchrzak** [comment]
It is patched indeed. Thank you :)
